CARDS-2348: New question type: postal address

veronikaslc opened the pull request:
- Utilises [react-google-autocomplete](https://www.npmjs.com/package/react-google-autocomplete) widget for the address suggestions
- Requires a Google API key configured either as a evironment var named `GOOGLE_APIKEY` or via the admin page `/content.html/admin/GoogleApiKey`. !!**Make sure to [follow the steps](https://help.stockist.co/article/43-verifying-your-google-maps-api-key) to enable the API key for Places services**!!.
- Can be tested running a test mode via creating an instance of the `Postal Address Test` Form
- New questions of the `address` type could be created via the questionnaire editor

veronikaslc committed:
CARDS-2348: New question type: postal address

veronikaslc committed:
CARDS-2348: New question type: postal address
Use plain javascript to get the key; Rendet text question for the case of no api key

veronikaslc committed:
CARDS-2348: New question type: postal address
Code cleanup

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/Question.jsx:
```diff
@@ -68,8 +68,8 @@ function Question (props) {
   }
-  if (questionDefinition.dataType == "phone") {
-	cardClasses.push(classes.phoneAnswer);
+  if (questionDefinition.dataType == "phone" || questionDefinition.dataType == "address") {
+    cardClasses.push(classes.phoneAnswer);
```
This is ugly.

veronikaslc replied:
True. Temporary solution till the discussion around that line is resolved in 2346. @marta- asked if any styling that is necessary could be done in the phone question component rather than top `Question` component. But I don't know ways to elegantly pass CSS styling up there. The container that defines overflow resides in `Question` component.

veronikaslc replied:
Resolved in 2346 by unsetting the overflow on Question container globally, rebased on latest 2346

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/AddressQuestion.jsx:
```diff
@@ -0,0 +1,159 @@
+const inputGlobalStyles = <GlobalStyles
+    styles={(theme) => ({
+      body: {
+        // to remove the "Powered by Google" logo from the bottom of the Google Map Autocomplete dropdown list
```
I'm not sure that's legally permitted. Someone should read the terms of use to check that we aren't required to display that.

veronikaslc replied:
The [other widget I tried](https://www.npmjs.com/package/react-google-places-autocomplete) with MIT license do not display it in the dropdown list so we should be fine.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/AddressQuestion.jsx:
```diff
@@ -0,0 +1,159 @@
+    if (googleApiKey) {
+      return [StyledAddressQuestion, 50];
+    } else {
+      return [StyledTextQuestion, 0];
```
I don't think this is needed.

```suggestion
```

sashaandjic commented:
ready fir test?

sdumitriu commented on test-resources/src/main/resources/SLING-INF/content/Questionnaires/AddressTest.xml:
```diff
@@ -0,0 +1,64 @@
+		<type>String</type>
+	</property>
+	<node>
+		<name>phone number</name>
```
```suggestion
		<name>postal address</name>
```

veronikaslc commented:
I'd wait for the comments first not to waste any time on testing in case the re-testing will be needed.

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/GoogleApiKeyServlet.java:
```diff
@@ -0,0 +1,68 @@
+ */
+@Component(service = { Servlet.class })
+@SlingServletResourceTypes(
+    resourceTypes = { "cards/ResourceHomepage" },
```
This is a system wide configuration that doesn't belong to a resource homepage. Can we try to apply it to `cards/Homepage`?

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/AddressQuestion.jsx:
```diff
@@ -0,0 +1,159 @@
+  const countries = questionDefinition.onlyCountries?.indexOf(",") > 0 ? questionDefinition.onlyCountries.replaceAll(" ", "").split(",") : questionDefinition.onlyCountries;
+
+  const { ref: materialRef } = usePlacesWidget({
+    apiKey: key,
```
```suggestion
    apiKey: googleApiKey,
```

sdumitriu committed:
CARDS-2348: New question type: postal address
Bugfix: wrong variable name

sdumitriu committed:
CARDS-2348: New question type: postal address
Use the root node as the servlet endpoint, since the API key is a global resource not bound to a specific resource type

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/AddressQuestion.jsx:
```diff
@@ -0,0 +1,159 @@
+    options: {
+      types: ["address"],
+      fields: ["formatted_address"],
+      componentRestrictions: { country: countries }
```
Without configuring this, an ugly stacktrace is logged:

> InvalidValueError: setComponentRestrictions: in property country: not a string; and not an Array

veronikaslc replied:
However that is a  localize suggestions already implemented from your top comment configurable from the questionnaire editor

veronikaslc replied:
Fixed in [0444fa4](https://github.com/data-team-uhn/cards/pull/1599/commits/0444fa40bc2c25da9c9fc819acdba05f7fed2298)

sdumitriu committed:
CARDS-2348: New question type: postal address
Simplify the component registration, there is no need to return the fallback component here

sdumitriu reviewed (commented):
Looks good so far. I've made a few changes.

One thing I don't like is that the suggestions are global, typing a number suggests things from all over the world, not nearby places. Someone should look through the props we could pass to localize suggestions.

veronikaslc committed:
CARDS-2348: New question type: postal address

fixed test form wording

Co-authored-by: Sergiu Dumitriu <sdumitriu@users.noreply.github.com>

veronikaslc committed:
CARDS-2348: New question type: postal address
Removed unused import

veronikaslc committed:
CARDS-2348: New question type: postal address
Avoid ugly console log if countries param is missing

veronikaslc committed:
CARDS-2348: New question type: postal address
Better countries param parsing

veronikaslc committed:
CARDS-2348: New question type: postal address

Test questionnaire adjustments

veronikaslc committed:
CARDS-2348: New question type: postal address
text is no longer required in question definitions

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/AddressQuestion.jsx:
```diff
@@ -0,0 +1,155 @@
+      <TextField
+        fullWidth
+        variant="standard"
+        placeholder="Enter the address"
```
I would either make this a helperText or remove it completely.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/AddressQuestion.jsx:
```diff
@@ -0,0 +1,155 @@
+      >
+      {inputGlobalStyles}
+      <TextField
+        fullWidth
```
I understand that it's full width because addresses can get quite long, but I don't like the length mismatch with other inputs in the form. Would making it multiline instead break anything?

marta- commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/googleApiKeyAdminPage.jsx:
```diff
@@ -0,0 +1,113 @@
+  }
+
+  return(
+    <AdminScreen title="Google API key configuration">
```
Could we use `AdminConfigScreen` instead, to have a consistent layout and buttons with other screens?

veronikaslc replied:
`AdminConfigScreen` serves general scenario of fetching `.json` configs which we we don't need here.

marta- replied:
Ok. Can copy and adapt the Save button behaviour (disable on successful savem re-enable on form change, etc)? Right now there's no feedback to the user on saving and it may look like it's not working.

marta- commented on test-resources/src/main/resources/SLING-INF/content/Questionnaires/AddressTest.xml:
```diff
@@ -0,0 +1,54 @@
+	<primaryNodeType>cards:Questionnaire</primaryNodeType>
+	<property>
+		<name>title</name>
+		<value>Postal Address Test</value>
```
Same small inconvenience like with "Telephone Numbers Test": `@name` and `title` mismatch so the entry looks out of place in the table that is sorted alphabetically by `@name`.
```suggestion
		<value>Address Test</value>
```

marta- commented on test-resources/src/main/resources/SLING-INF/content/Questionnaires/AddressTest.xml:
```diff
@@ -0,0 +1,54 @@
+		<type>String</type>
+	</property>
+	<node>
+		<name>postal address</name>
```
```suggestion
		<name>postal_address</name>
```

marta- commented on test-resources/src/main/resources/SLING-INF/content/Questionnaires/AddressTest.xml:
```diff
@@ -0,0 +1,54 @@
+		<primaryNodeType>cards:Question</primaryNodeType>
+		<property>
+			<name>text</name>
+			<value>postal address</value>
```
```suggestion
			<value>Postal address</value>
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/Question.json:
```diff
@@ -453,6 +453,10 @@
         "onlyCountries": "string",
         "regions": "string"
+      },
+      "address" : {
+        "minAnswers": {"0":{}, "1":{}},
+        "onlyCountries": "string"
```
I feel like there should be a way to provide a default for question definition properties somewhere. The AddressQuestion component assumes maxAnswers=1 and displays one text field only. But if that component is not active because the google api key is missing, we fall back to the completely generic text component that by default is multivalued, which is inconsistent.

veronikaslc replied:
We can specify a `defaultMaxAnswers` in `.cnd` file, would that work?

veronikaslc committed:
CARDS-2348: New question type: postal address

Better wording in test questionnaire

Co-authored-by: Marta Girdea <marta@phenotips.org>

veronikaslc committed:
CARDS-2348: New question type: postal address
Use multiline for textinput

veronikaslc committed:
CARDS-2348: New question type: postal address
Better error and state handling for the Save button

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/AddressQuestion.jsx:
```diff
@@ -0,0 +1,155 @@
+      {...props}
+      >
+      {inputGlobalStyles}
+      <TextField
```
This is now narrower than other question text fields.

marta- replied:
Fixed in ff587bc

marta- commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/googleApiKeyAdminPage.jsx:
```diff
@@ -0,0 +1,126 @@
+  return(
+    <AdminScreen title="Google API key configuration">
+      <Grid container direction="column" spacing={6} justifyContent="space-around">
+        { !googleSystemApiKey &&
+         <Grid item className={classes.noKeyInfo}>
+           <Typography>Your system does not have a Google API Key configured.</Typography>
+           <Typography>Without an API key, you cannot access Google services such as address autocomplete.</Typography>
+         </Grid> }
```
I'm not a fan of how the previously empty input jumps up once the key is submitted because of the disappearance of this message.

marta- replied:
Addressed in 1450084

marta- commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/googleApiKeyAdminPage.jsx:
```diff
@@ -0,0 +1,126 @@
+            spacing={2}
+          >
+            <Grid item xs={10}>
+              <TextField
```
The text field is disproportionately tall compared to the button.

marta- replied:
Addressed in 407ee0e

marta- committed:
CARDS-2348: New question type: postal address
Make the address text field the same size as the default text field

marta- committed:
CARDS-2348: New question type: postal address
Better intro message for the API key config

marta- committed:
CARDS-2348: New question type: postal address
Made text field and button have considtent sizes on the  API key config screen

marta- committed:
CARDS-2348: New question type: postal address
Minor refactoring

marta- commented:
Would it make sense to move everything related to configuring the API key (i.e. servlet, admin page) outside of data-entry and into its own module? @sdumitriu

sdumitriu commented:
Yes it would. Is the key specific to the Maps API or is it for all Google services?

veronikaslc commented:
The key is one for any service you would turn on for this key in the Google settings.

marta- commented:
Ok, let's move them to a new module `google-api`.

veronikaslc committed:
CARDS-2348: New question type: postal address
Separated google API related functionality to a stand-alone component

marta- commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/Question.json:
```diff
@@ -453,6 +453,10 @@
+      },
+      "address" : {
+        "minAnswers": {"0":{}, "1":{}},
+        "onlyCountries": "string"
```
We should add a hint for `onlyCountries`:

```
When address autocomplete is enabled by providing a Google API key, suggestions will be restricted to addresses from the countries listed here. Example value: `ca, us`.\n\nYou can obtain a Google API key at [put the URL here] and add it to **Administration** > **Google API key configuration**.
```

marta- replied:
Oh. We can't add a hint that is different from the one provided for phone/onlyCountries.

marta- replied:
I'm inclined to suggest changing the name of the `onlyCountries` property for `address` (for example to `countries`) so we can have separate hints for `address` and `phone`. Thoughts?

veronikaslc committed:
CARDS-2348: New question type: postal address
Added hint for the countries param for the google autocomplete widget

marta- commented on modules/google-api/src/main/java/io/uhndata/cards/google/GoogleApiKeyManager.java:
```diff
@@ -0,0 +1,49 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package io.uhndata.cards.google;
```
Not a fan of the package name. `io.uhhdata.cards.googleapi`? @sdumitriu is there any naming convention we should be following here?

sdumitriu replied:
`io.uhndata.cards.googleapis`, with an s at the end.

marta- replied:
Thanks. Should the module name have an `s` at the end as well? Currently it's `google-api`.

sdumitriu replied:
Yes.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/Question-hints.json:
```diff
@@ -5,6 +5,7 @@
   "regions": "In the country dropdown, show only countries from certain regions or subregions, specified as a comma-separated list. For example:\n*  Regions - `america, europe, asia, oceania, africa`);\n* Subregions -`north-america, south-america, central-america, carribean, eu-union, ex-ussr, ex-yugos, baltic, middle-east, north-africa`.",
-  "onlyCountries": "Explicit list of comma-separated country codes to be made available in the dropdown, for example `es, fr, it`."
+  "onlyCountries": "Explicit list of comma-separated country codes to be made available in the dropdown, for example `es, fr, it`.",
+  "countries": "When address autocomplete is enabled by providing a Google API key, suggestions will be restricted to addresses from the countries listed here. Example value: `ca, us`.\n\nYou can obtain a Google API key at [https://developers.google.com/maps/documentation/javascript/get-api-key] and add it to **Administration** > **Google API key configuration**."
```
Changed my mind after testing...

Let's simplify this hint:
```suggestion
  "countries": "When address autocomplete is enabled by providing a Google API key, suggestions will be restricted to addresses from the countries listed here. Example value: `ca, us`.\n\nYou can set up the Google API key in **Administration** > **Google API key configuration**."
```

marta- commented on modules/google-api/src/main/frontend/src/googleApiKeyAdminPage.jsx:
```diff
@@ -0,0 +1,115 @@
+    <AdminScreen title="Google API key configuration">
+      <Grid container direction="column" spacing={5} justifyContent="space-around">
+        <Grid item>
+          <Typography>An API key enables access to Google services such as address autocomplete.</Typography>
```
... and let's make this message more helpful:
```suggestion
          <FormattedText>A Google API key enables access to Google services such as address autocomplete. You can obtain an API key at https://developers.google.com/maps/documentation/javascript/get-api-key  .</FormattedText>
```

marta- commented on modules/google-api/src/main/frontend/src/googleApiKeyAdminPage.jsx:
```diff
@@ -0,0 +1,115 @@
+  Button,
+  Grid,
+  TextField,
+  Typography,
```
`Typography` will no longer be used, and we must import `FormattedText`.

veronikaslc committed:
CARDS-2348: New question type: postal address

Better address datatype hint and Google API Key admin screen intro message.

marta- commented on modules/google-api/src/main/frontend/src/googleApiKeyAdminPage.jsx:
```diff
@@ -24,7 +24,7 @@ import {
   Grid,
   TextField,
-  Typography,
+  FormattedText,
```
FormattedText is not a Material component, it's one of ours

marta- commented:
> Not a fan of the package name. `io.uhhdata.cards.googleapi`? @sdumitriu is there any naming convention we should be following here?

This is the only comment I have left before approval, let's decide on the package naming @veronikaslc @sdumitriu .

marta- commented on modules/google-api/src/main/resources/SLING-INF/content/Extensions/AdminDashboard/AdminGoogleApiKey.json:
```diff
@@ -0,0 +1,9 @@
+{
+  "jcr:primaryType": "cards:Extension",
+  "cards:extensionPointId": "cards/coreUI/admindashboard/entry",
+  "cards:extensionName": "Google Api Key",
```
```suggestion
  "cards:extensionName": "Google API Key",
```

marta- committed:
CARDS-2348: New question type: postal address

Fix the name of the Google API Key admin entry

veronikaslc committed:
CARDS-2348: New question type: postal address
Better name for the bundle

sashaandjic commented:
I see the new `address` type. However it does not do anything for me. What am I missing?
![image](https://github.com/data-team-uhn/cards/assets/23243244/4e4833c1-4e4b-409c-abb2-d63903231970)
I put the API key in ADMINSTRATION/Google API key config page.

veronikaslc commented:
Honestly it's hard to tell like that from the scrinshot alone. What do you exactly do? Are there any console errors? If its not video there is not much to conclude.

sashaandjic commented:
Once I did export GOGGLE_APIKEY=xxxxxxx I got different behaviour.

veronikaslc commented:
Just made a fresh build and tested the admin setting of the key - worked for me.

sashaandjic reviewed (approved):
tested - awesome!!!

veronikaslc committed:
CARDS-2348: New question type: postal address
Removed unused var

veronikaslc committed:
CARDS-2348: New question type: postal address
The AddressQuestion won't fetch new key untill the page is reloaded

veronikaslc committed:
CARDS-2348: New question type: postal address
Fall back to the TextQuestion if Google API authentication fails; Set input to the formatted selection if made;

sashaandjic reviewed (approved):
re-tested

I no longer need to refresh once I enter a GOOGLE_API key in admin

veronikaslc commented:
Rebased on the latest `dev`.

sdumitriu commented on modules/data-model/forms/api/src/main/resources/SLING-INF/nodetypes/forms.cnd:
```diff
@@ -153,6 +153,7 @@
   // - computed: a derived value automatically computed based on a formula applied on other values in the questionaire
   // - pedigree: a pedigree drawing
   // - phone: a phone or fax number
+  // - address: 
```
Missing description.

sdumitriu committed:
CARDS-2348: New question type: postal address
Add a new configuration for the question, searchPlacesAround, to bias the results towards an area

sdumitriu committed:
CARDS-2348: New question type: postal address
Fix the parent version number

sdumitriu committed:
CARDS-2348: New question type: postal address
Improve documentation

sdumitriu reviewed (approved):
I've added support for a `searchPlacesAround` property to bias results around a place, and I can now say that I'm almost happy with the results.

One remaining problem is that we're using a textarea instead of an input, and the Google API is complaining about that; the API does say that it **must** be an `HtmlInputElement`, so we're wrong, although it seems to work fine. I would like to be compliant with the API requirements, but I also like that the answer grows vertically to show the full answer.

Once, while testing, it somehow stopped selecting the suggested address when clicking on it, although using up-down arrows and enter to select it still worked.

I'm not entirely happy with the fact that sometimes the selected address is not the one that was displayed, for example if I type `666B Bay Street`, it does suggest `666b Bay Street`, but when I click on it it becomes `666 Bay St.`. I can manually correct it to what I wanted, but I hope users will not accidentally enter the wrong address. Technically, that is the correct behavior since we're using a **place** suggestion, not an **address** suggestion, so the workflow behind the scenes is:
- get the user's input
- suggest addresses based on the input
- when an address is selected, we get the place for that address
- we get the place's address and use it
I'd like to see if it's possible to request an address suggestion directly instead of a place suggestion, but not now.

marta- commented:
Todo: update the test questionnaire to showcase the newly introduced `searchPlacesAround` setting @sdumitriu

marta- commented:
@sdumitriu we would need this merged for sparc. Are you able to update the test questionnaire and then merge?

sdumitriu committed:
CARDS-2348: New question type: postal address
More test questions
